engine/access/subscription/data_provider.go

```diff
@@ -0,0 +1,71 @@
+package subscription
+
+import (
+	"context"
+	"fmt"
+)
+
+// DataProvider represents a source of sequential items to be streamed to a client.
+//
+// TODO: It can and should operate on generic type  rather than `any`. IT'd require lots of refactoring,
+// so it should be done in a PR with the subscription package refactoring.
+type DataProvider interface {
+	// NextData retrieves the next sequential item from the data source.
+	//
+	// - Retrieves the next available item in sequence and returns it as the first result.
+	// - If no further item is currently available yet (e.g. the next block/data is not ready), it
+	//   should return a non-nil error that satisfies [ErrBlockNotReady]. The streamer interprets this
+	//   as a normal "no more right now" condition and will pause until new data is broadcast.
+	// - To signal a terminal end of the stream (no more data will ever be available for this
+	//   subscription), return a non-nil error that satisfies [ErrEndOfData]. The streamer will close
+	//   the subscription gracefully.
+	// - Must observe the provided context: if the context is canceled or its deadline is exceeded,
+	//   return the respective context error.
```

Comment:
I think this is implied by accepting a context and documenting the expected error. you can omit it